// ibm-cloud-sdk-core/lib/sdk-test-helpers.browser.js
"use strict";
/* eslint-disable @typescript-eslint/no-unused-vars */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Copyright 2021 IBM Corp. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// Dummy for browser
function checkUrlAndMethod(options, url, method) { }
exports.checkUrlAndMethod = checkUrlAndMethod;
function checkMediaHeaders(createRequestMock, accept, contentType) { }
exports.checkMediaHeaders = checkMediaHeaders;
function checkUserHeader(createRequestMock, userHeaderName, userHeaderValue) { }
exports.checkUserHeader = checkUserHeader;
function checkForSuccessfulExecution(createRequestMock) { }
exports.checkForSuccessfulExecution = checkForSuccessfulExecution;
function getOptions(createRequestMock) { }
exports.getOptions = getOptions;
function expectToBePromise(sdkPromise) { }
exports.expectToBePromise = expectToBePromise;
